runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java

```diff
@@ -0,0 +1,170 @@
+      throw new IllegalArgumentException("The given amount of memory amounted to less than one chunk.");
+    }
+    this.chunkSize = chunkSizeKb * 1024;
+    this.memoryPool = new MemoryPool((int) maxNumChunks, this.chunkSize);
+  }
+
+  /**
+   * Returns a single {@link MemoryChunk} from {@link MemoryPool}.
+   *
+   * @return a MemoryChunk
+   * @throws MemoryAllocationException if fails to allocate MemoryChunk.
+   */
+  public MemoryChunk allocateChunk() throws MemoryAllocationException {
+    return memoryPool.requestChunkFromPool();
+  }
+
+  /**
+   * Returns all the MemoryChunks in the given List of MemoryChunks.
+   *
+   * @param target  the list of MemoryChunks to be returned to the memory pool.
+   */
+  public void returnChunksToPool(final List<MemoryChunk> target) {
+    for (final MemoryChunk chunk: target) {
+      memoryPool.returnChunkToPool(chunk);
+    }
+  }
+
+  /**
+   * Returns the chunk size of the memory pool.
+   *
+   * @return the chunk size in bytes.
+   */
+  public int getChunkSize() {
+    return chunkSize;
+  }
+
+  @VisibleForTesting
+  /**
+   * Returns the number of chunks in the pool.
+   */
+  public int returnPoolSize() {
+    return memoryPool.returnPoolSize();
+  }
+
+  /**
+   * Memory pool that utilizes off-heap memory.
+   * Supports pre-allocation of memory according to user specification.
+   */
+  @ThreadSafe
+  private class MemoryPool {
+
+    private final ConcurrentLinkedQueue<ByteBuffer> pool;
+    private final int chunkSize;
+    private long maxNumChunks;
+    private long numChunks;
+
+    MemoryPool(final long maxNumChunks, final int chunkSize) {
+      this.chunkSize = chunkSize;
+      this.pool = new ConcurrentLinkedQueue<>();
+      this.maxNumChunks = maxNumChunks;
+    }
+
+    synchronized MemoryChunk allocateNewChunk() throws MemoryAllocationException {
+      if (maxNumChunks <= numChunks) {
+        throw new MemoryAllocationException("Exceeded maximum off-heap memory");
+      }
+      ByteBuffer memory = ByteBuffer.allocateDirect(chunkSize);
+      numChunks += 1;
+      return new MemoryChunk(memory);
+    }
+
+    MemoryChunk requestChunkFromPool() throws MemoryAllocationException {
+      try {
+        if (pool.isEmpty()) {
+          return allocateNewChunk();
+        } else {
+          ByteBuffer buf = pool.remove();
+          return new MemoryChunk(buf);
+        }
+      } catch (final NoSuchElementException e) {
+        throw new MemoryAllocationException("Pool empty: Failed to retrieve MemoryChunk");
+      } catch (final OutOfMemoryError e) {
+        throw new MemoryAllocationException("Memory allocation failed due to lack of memory");
+      }
+    }
+
+    /**
+     * Returns MemoryChunk back to memory pool.
+     *
+     * @param chunk the target MemoryChunk to be returned to the pool.
+     */
+    void returnChunkToPool(final MemoryChunk chunk) {
+      ByteBuffer buf = chunk.getBuffer();
+      buf.clear();
+      pool.add(buf);
+      chunk.release();
```

Comment:
I think L159-L160 here should go inside `chunk.release()` to make things consistent. I also think `chunk.release()` should be invoked prior to `pool.add(buf)`, to ensure that a clean chunk is added back to the pool. With multi-threaded access, the chunk can be used **before** the invocation of `chunk.release()`, which may affect correctness.